### PR TITLE
PROD RING pipeline support

### PR DESCRIPTION
**Pull request by VaclavK**

### CODE REVIEW

**Comment by VaclavK** [comment]
one thing I realised here is that the version concept for sierra is not only code but also code + underlying artefacts used as templates by code (e.g. VSTS build or release templates) so when the template changes (but code does not) we will have to recreate specific artefacts anyway

File: src/Sierra.Model/Sierra.Model.csproj [hunk 1]
```diff
@@ -7,6 +7,14 @@
     <LangVersion>latest</LangVersion>
   </PropertyGroup>
 
+  <ItemGroup>
```
Comment (arturek) [inline src/Sierra.Model/Sierra.Model.csproj]: not needed

File: src/Sierra.Common/ReleaseHttpClient2Extensions.cs [hunk 2]
```diff
@@ -71,5 +69,31 @@ public static async Task DeleteReleaseDefinitionIfFExists(this ReleaseHttpClient
...
+        /// <param name="targetProject">target project</param>
+        /// <param name="definitionName">definition name to load</param>
+        /// <returns></returns>
+        public static async Task<ReleaseDefinition> LoadDefinitionByNameIfExists(this ReleaseHttpClient2 client, string targetProject,
+            string definitionName)
+        {
+            //if ring based, it may already exists (since shared by tenants), try to load it
+            var definition = (await client.GetReleaseDefinitionsAsync(
+                    targetProject,
+                    definitionName, isExactNameMatch: true))
```
Comment (arturek) [inline src/Sierra.Common/ReleaseHttpClient2Extensions.cs]: inconsistent formatting (newline)

File: src/Sierra.Actor/ReleaseDefinitionActor.cs [hunk 3]
```diff
@@ -64,16 +66,98 @@ public class ReleaseDefinitionActor : SierraActor<VstsReleaseDefinition>, IRelea
...
+
+            return model;
+        }
+
+        private void CustomizeRingPipeline(VstsReleaseDefinition model, ReleaseDefinition pipeline)
+        {
+            var variableName = GetTenantSizePipelineVariableName(model.TenantSize);
+
+            if (!pipeline.Variables.ContainsKey(variableName))
+                throw new Exception($"Ring template #{pipeline.Id} does not contain expected variable {variableName}");
```
[the remark this answers is not in the document]
Comment (VaclavK) [reply src/Sierra.Actor/ReleaseDefinitionActor.cs]: yeah I did not want to start creating new ones prior to the exception handling refactor - not knowing what they should look like/base types etc... but I can at least introduce them

File: src/Sierra.Model/Sierra.Model.csproj (hunk 1 above)
Comment (VaclavK) [reply src/Sierra.Model/Sierra.Model.csproj]: thanks.. I &$£# how they treat removed migrations...

File: src/Tests/Sierra.Actor.Tests/ReleaseDefinitionActorTests.cs [hunk 4]
```diff
@@ -116,6 +117,151 @@ public async Task AddStandardTest()
...
+                    resp1.VstsReleaseDefinitionId);
+                vstsRel.Should().NotBeNull();
+                //check variable containing definition
+                vstsRel.Variables["SmallTenants"].Value.Should().Be("L2TNT2#11111;L2TNT#11111");
+                pipelineId = resp1.VstsReleaseDefinitionId;
+            }
+            finally
+            {
+                await cl.PostJsonToActor(Fixture.TestMiddlewareUri,
+                    "ReleaseDefinition", "Remove",
```
Comment (dguerin) [inline src/Tests/Sierra.Actor.Tests/ReleaseDefinitionActorTests.cs]: inconsistent formatting (newline)

File: src/Tests/Sierra.Actor.Tests/ReleaseDefinitionActorTests.cs [hunk 5]
```diff
@@ -116,6 +117,151 @@ public async Task AddStandardTest()
...
+                    releaseDefinition1);
+                //check tenant definition gone
+
+                var vstsRel = await releaseClient.GetReleaseDefinitionAsync(vstsConfig.VstsTargetProjectId,
+                    pipelineId);
+                vstsRel.Variables["SmallTenants"].Value.Should().Be("L2TNT2#11111");
+
+                await cl.PostJsonToActor(Fixture.TestMiddlewareUri,
+                    "ReleaseDefinition", "Remove",
```
Comment (dguerin) [inline src/Tests/Sierra.Actor.Tests/ReleaseDefinitionActorTests.cs]: inconsistent formatting (newline)

File: src/Tests/Sierra.Actor.Tests/ReleaseDefinitionActorTests.cs [hunk 6]
```diff
@@ -116,6 +117,151 @@ public async Task AddStandardTest()
...
+            }
+            finally
+            {
+                await cl.PostJsonToActor(Fixture.TestMiddlewareUri,
+                    "ReleaseDefinition", "Remove",
+                    releaseDefinition1);
+                //check tenant definition gone
+
+                var vstsRel = await releaseClient.GetReleaseDefinitionAsync(vstsConfig.VstsTargetProjectId,
+                    pipelineId);
```
Comment (dguerin) [inline src/Tests/Sierra.Actor.Tests/ReleaseDefinitionActorTests.cs]: inconsistent formatting (newline)

**Pull request merged**